ruby/ql/src/queries/variables/UninitializedLocal.ql

```diff
@@ -5,20 +5,71 @@
  * @kind problem
  * @problem.severity error
  * @id rb/uninitialized-local-variable
- * @tags reliability
+ * @tags quality
+ *       reliability
  *       correctness
- * @precision low
+ * @precision medium
  */
 
 import codeql.ruby.AST
 import codeql.ruby.dataflow.SSA
+private import codeql.ruby.dataflow.internal.DataFlowPublic
+import codeql.ruby.controlflow.internal.Guards as Guards
+import codeql.ruby.controlflow.CfgNodes
+
+predicate isInBooleanContext(Expr e) {
+  e = any(ConditionalExpr c).getCondition()
+  or
+  e = any(ConditionalLoop l).getCondition()
+  or
+  e = any(LogicalAndExpr n).getAnOperand()
+  or
+  e = any(LogicalOrExpr n).getAnOperand()
+  or
+  e = any(NotExpr n).getOperand()
+}
+
+predicate isGuarded(LocalVariableReadAccess read) {
+  exists(AstCfgNode guard, boolean branch |
+    Guards::guardControlsBlock(guard, read.getAControlFlowNode().getBasicBlock(), branch)
+  |
+    // guard is `var`
+    guard.getAstNode() = read.getVariable().getAnAccess() and
+    branch = true
+    or
+    // guard is `!var`
+    guard.getAstNode().(NotExpr).getOperand() = read.getVariable().getAnAccess() and
+    branch = false
+    or
```

Comment:
This disjunct can be removed; the CFG automatically handles negation.

Reply:
Ah, nice!